File: app/src/main/java/com/lxj/easyadapter/sample/MainActivity.java
package com.lxj.easyadapter.sample;

import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.lxj.easyadapter.CommonAdapter;
import com.lxj.easyadapter.ItemViewDelegate;
import com.lxj.easyadapter.MultiItemTypeAdapter;
import com.lxj.easyadapter.ViewHolder;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    RecyclerView recyclerView;

    List<User> userList = new ArrayList<>();
    private CommonAdapter<User> adapter;
    private MultiItemTypeAdapter multiItemTypeAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        recyclerView = findViewById(R.id.rv);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        //prepare data
        for (int i = 0; i < 20; i++) {
            userList.add(new User("本杰明 - ", i * 2));
        }

        testHeader();
//        testMultiItem();
    }

    void testHeader() {
        adapter = new CommonAdapter<User>(R.layout.item, userList) {
            @Override
            protected void bind(ViewHolder holder, User user, int position) {
                holder.setText(R.id.tv_name, "name: " + user.name + position)
                        .setText(R.id.tv_age, "age: " + user.age);
            }
        };
        adapter.setOnItemClickListener(new MultiItemTypeAdapter.SimpleOnItemClickListener() {
            @Override
            public void onItemClick(View view, RecyclerView.ViewHolder holder, int position) {
                super.onItemClick(view, holder, position);
                Toast.makeText(MainActivity.this, "item - " + position, Toast.LENGTH_SHORT).show();
            }
        });

        adapter.addHeaderView(createView("Header - 1"));
        adapter.addHeaderView(createView("Header - 2"));
        adapter.addFootView(createView("我是Footer - 1"));
        adapter.addFootView(createView("我是Footer - 2"));
        recyclerView.setAdapter(adapter);
    }

    TextView createView(String text) {
        final TextView textView2 = new TextView(this);
        textView2.setPadding(80, 80, 80, 80);
        textView2.setText(text);
        textView2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Toast.makeText(MainActivity.this, textView2.getText().toString(), Toast.LENGTH_SHORT).show();
                userList.add( new User("刘小姐", 33));
                if(multiItemTypeAdapter!=null)multiItemTypeAdapter.notifyDataSetChanged();
                if(adapter!=null)adapter.notifyDataSetChanged();
            }
        });
        return textView2;
    }

    void testMultiItem() {

        multiItemTypeAdapter = new MultiItemTypeAdapter<User>(userList)
                .addItemViewDelegate(new OneDelegate())
                .addItemViewDelegate(new TwoDelegate());
        multiItemTypeAdapter.addHeaderView(createView("Multi Header view1111"));
        multiItemTypeAdapter.addHeaderView(createView("Multi Header view22222"));
        multiItemTypeAdapter.addFootView(createView("Multi Footer view"));
        multiItemTypeAdapter.setOnItemClickListener(new MultiItemTypeAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(@NonNull View view, @NonNull RecyclerView.ViewHolder holder, int position) {
                Toast.makeText(MainActivity.this, "position: " + position, Toast.LENGTH_SHORT).show();
            }

            @Override
            public boolean onItemLongClick(@NonNull View view, @NonNull RecyclerView.ViewHolder holder, int position) {
                return false;
            }
        });
        recyclerView.setAdapter(multiItemTypeAdapter);
    }

    class OneDelegate implements ItemViewDelegate<User> {

        @Override
        public int getLayoutId() {
            return android.R.layout.simple_list_item_1;
        }

        @Override
        public boolean isForViewType(@NonNull User item, int position) {
            return position % 2 != 0;
        }

        @Override
        public void bind(@NonNull ViewHolder holder, @NonNull User user, int position) {
            holder.setText(android.R.id.text1, "name: " + user.name + " - "+ position);
        }
    }

    class TwoDelegate implements ItemViewDelegate<User> {

        @Override
        public int getLayoutId() {
            return android.R.layout.simple_list_item_1;
        }

        @Override
        public boolean isForViewType(@NonNull User item, int position) {
            return position % 2 == 0;
        }

        @Override
        public void bind(@NonNull ViewHolder holder, @NonNull User user, int position) {
            holder.setText(android.R.id.text1, "age: " + user.age)
                    .getView(android.R.id.text1).setBackgroundColor(Color.RED);
        }
    }
}

File: app/src/main/java/com/lxj/easyadapter/sample/MainActivity.kt
package com.lxj.easyadapter.sample

import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.blankj.utilcode.util.CloneUtils
import com.blankj.utilcode.util.GsonUtils
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.request.RequestOptions
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.lxj.easyadapter.*

import kotlinx.android.synthetic.main.activity_main.*
import java.lang.reflect.Type
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList
import com.bumptech.glide.request.transition.DrawableCrossFadeFactory




class MainActivity : AppCompatActivity() {

    internal var userList: MutableList<User> = ArrayList()
    private var adapter: EasyAdapter<User>? = null
    private var multiItemTypeAdapter: MultiItemTypeAdapter<User>? = null
    val url = "https://img1.baidu.com/it/u=1925715390,133119052&fm=253&fmt=auto&app=138&f=JPEG?w=400&h=400"
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        userList.add(User(name = "盛大"))
        userList.add(User(name = "都是"))
        userList.add(User(name = "CEAD"))
        userList.add(User(name = "大图"))
        add.setOnClickListener {
            val old = deepCopy()
            val range = (0 until userList.size)
            userList.add(if(range.isEmpty()) 0 else range.random(), User(id = UUID.randomUUID().toString(), name = "随机添加 - ${(0..1000).random()}" ))
            DiffUtil.calculateDiff(UserDiffCallback(old, userList)).dispatchUpdatesTo(recyclerView.adapter!!)
        }
        del.setOnClickListener {
            if(userList.isNullOrEmpty()) return@setOnClickListener
            val range = (0 until userList.size)
            val old = deepCopy()
            userList.removeAt(range.random())
            DiffUtil.calculateDiff(UserDiffCallback(old, userList)).dispatchUpdatesTo(recyclerView.adapter!!)
        }
        update.setOnClickListener {
            if(userList.isNullOrEmpty()) return@setOnClickListener
            val index = (0 until userList.size).random()
            val old = deepCopy()
            userList[index].name += " - 随机更新"
            DiffUtil.calculateDiff(UserDiffCallback(old, userList)).dispatchUpdatesTo(recyclerView.adapter!!)
        }
        replace.setOnClickListener {
            if(userList.isNullOrEmpty()) return@setOnClickListener
            val index = (0 until userList.size).random()
            val old = deepCopy()
            userList[index] = User(name = "我是随机替换的")
            DiffUtil.calculateDiff(UserDiffCallback(old, userList)).dispatchUpdatesTo(recyclerView.adapter!!)
        }
        move.setOnClickListener {
            if(userList.isNullOrEmpty()) return@setOnClickListener
            val old = deepCopy()
            userList.reverse()
            DiffUtil.calculateDiff(UserDiffCallback(old, userList)).dispatchUpdatesTo(recyclerView.adapter!!)
        }
        recyclerView.layoutManager = LinearLayoutManager(this)

        //prepare data
//        for (i in 0..6) {
//            userList.add(User("本杰明 - $i", i * 2, i))
//        }

//        testHeader()
        testMultiItem()
    }

    fun deepCopy(): List<User>{
        val json = GsonBuilder().create().toJson(userList)
        return GsonBuilder().create().fromJson<List<User>>(json, object : TypeToken<List<User>>() {}.type)
    }

    private fun testHeader() {
        adapter = object : EasyAdapter<User>(userList, R.layout.item) {
            override fun bind(holder: ViewHolder, user: User, position: Int) {
                with(holder) {
                    setText(R.id.tv_name, "name: " + user.name )
                    setText(R.id.tv_age, "age: " + user.age)
                }
            }

        }.apply {
            setOnItemClickListener(object : MultiItemTypeAdapter.SimpleOnItemClickListener() {
                override fun onItemClick(view: View, holder: RecyclerView.ViewHolder, position: Int) {
                    super.onItemClick(view, holder, position)
                    Toast.makeText(this@MainActivity, "position - $position", Toast.LENGTH_SHORT).show()
                    userList.removeAt(position)
//                    notifyDataSetChanged()
                    notifyItemRemoved(position + headersCount)
                }
            })
            addHeaderView(createView("Header - 1，点我在头部添加一条数据"))
            addHeaderView(createView("Header - 2，点我在头部添加一条数据"))
            addFootView(createView("Footer - 1，点我在末尾添加一条数据", true))
            addFootView(createView("Footer - 2，点我在末尾添加一条数据", true))
            recyclerView.adapter = this
        }
    }

    private fun createView(text: String, isFooter: Boolean = false): TextView {
        val textView = TextView(this)
        textView.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        textView.setPadding(80, 80, 80, 80)
        textView.text = text
        textView.setOnClickListener {
            Toast.makeText(this@MainActivity, textView.text.toString(), Toast.LENGTH_SHORT).show()
            val user = User()
            user.name = "随机添加的名字 - ${System.currentTimeMillis()}"
            user.age = System.currentTimeMillis().toInt()
            if(isFooter){
                //尾部添加
                userList.add(user)
            }else{
                userList.add(0, user)
            }
            multiItemTypeAdapter?.apply { notifyItemInserted(if (isFooter) (userList.size + headersCount) else headersCount) }
            adapter?.apply { notifyItemInserted(if (isFooter) (userList.size + headersCount) else headersCount) }
        }
        return textView
    }

    internal fun testMultiItem() {
        multiItemTypeAdapter = MultiItemTypeAdapter<User>(userList)
                .apply {
                    addItemDelegate(OneDelegate())
                    addItemDelegate(TwoDelegate())
//                    addHeaderView(createView("Multi Header view1111"))
//                    addHeaderView(createView("Multi Header view22222"))
//                    addFootView(createView("Multi Footer view"))
                    setOnItemClickListener(object : MultiItemTypeAdapter.OnItemClickListener {
                        override fun onItemClick(view: View, holder: RecyclerView.ViewHolder, position: Int) {
                            Toast.makeText(this@MainActivity, "position: $position", Toast.LENGTH_SHORT).show()
                        }

                        override fun onItemLongClick(view: View, holder: RecyclerView.ViewHolder, position: Int): Boolean {
                            return false
                        }
                    })
                    recyclerView.adapter = this
                }
    }

    internal inner class OneDelegate : ItemDelegate<User> {

        override fun isThisType(item: User, position: Int): Boolean {
            return position < 2
        }
        override fun bind(holder: ViewHolder, user: User, position: Int) {
            holder.setText(R.id.name, user.name)
            Glide.with(this@MainActivity)
                .load(url).transition(DrawableTransitionOptions.withCrossFade(1000))
                .diskCacheStrategy(DiskCacheStrategy.NONE)
                .skipMemoryCache(true)
                .placeholder(R.mipmap.ic_launcher_round)
                .into(holder.getView<ImageView>(R.id.avatar))
        }

        //布局更新
        override fun bindWithPayloads(holder: ViewHolder, t: User,
            position: Int, payloads: List<Any>) {
            if(payloads.isNullOrEmpty()) return
            val bundle = payloads[0] as Bundle
            val name = bundle.getString("name") ?: ""
            if(!name.isNullOrEmpty()){
                holder.setText(R.id.name, name)
            }
        }

        override fun getLayoutId(): Int = R.layout.adapter_one
    }

    internal inner class TwoDelegate : ItemDelegate<User> {
        override fun getLayoutId(): Int  = android.R.layout.simple_list_item_1

        override fun isThisType(item: User, position: Int): Boolean {
            return position >=2
        }
        override fun bind(holder: ViewHolder, user: User, position: Int) {
            holder.setText(android.R.id.text1, "age: " + user.age)
            holder.getView<View>(android.R.id.text1).setBackgroundColor(Color.RED)
        }
    }
}
